src/metrics/mod.rs

```diff
@@ -0,0 +1,100 @@
+#[macro_use]
+mod macros;
+
+use self::macros::MetricFromOpts;
+use crate::db::Pool;
+use crate::BuildQueue;
+use failure::Error;
+use prometheus::proto::MetricFamily;
+
+load_metric_type!(IntGauge as single);
+load_metric_type!(IntCounter as single);
+load_metric_type!(IntCounterVec as vec);
+load_metric_type!(HistogramVec as vec);
+
+metrics! {
+    pub struct Metrics {
+        /// Number of crates in the build queue
+        queued_crates_count: IntGauge,
+        /// Number of crates in the build queue that have a positive priority
+        prioritized_crates_count: IntGauge,
+        /// Number of crates that failed to build
+        failed_crates_count: IntGauge,
+
+        /// The number of idle database connections
+        idle_db_connections: IntGauge,
+        /// The number of used database connections
+        used_db_connections: IntGauge,
+        /// The maximum number of database connections
+        max_db_connections: IntGauge,
+        /// Number of attempted and failed connections to the database
+        pub(crate) failed_db_connections: IntCounter,
+
+        /// The number of currently opened file descriptors
+        open_file_descriptors: IntGauge,
+        /// The number of threads being used by docs.rs
+        running_threads: IntGauge,
```

Comment:
Need to add `$(#[$meta:meta])*` captures to the macro so that these two metrics can be cfg'd out for `#[cfg(target_os = "linux")]`
```suggestion
        /// The number of currently opened file descriptors
        #[cfg(target_os = "linux")]
        open_file_descriptors: IntGauge,
        /// The number of threads being used by docs.rs
        #[cfg(target_os = "linux")]
        running_threads: IntGauge,
```

Reply:
Why do you want to cfg out the metrics? We can just avoid setting them.

Reply:
If they aren't cfg'd out then Windows/MacOS users will receive dead code warnings (and if isn't changed to `#[cfg(not(target_os = "linux"))]` then MacOS users won't be able to compile)

Reply:
Makes sense! `#[cfg(linux)]` also works btw.